### ISSUE TITLE
Unable to generate moves in some FEN

### ISSUE THREAD

**Issue by govind-maheshwari2** [alexobviously/bishop#74]
Bishop unable to make move in `k7/5p2/8/6P1/8/8/8/K7 b` FEN
This issue only occur when running locally in web
Error Log:
```
Error: RangeError (index): Index out of range: index must not be negative: -1
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:49      throw_
dart-sdk/lib/_internal/js_dev_runtime/private/js_array.dart 591:7                 _get]
packages/sena_core/src/pgn_parser/utils/chess.dart 1246:38                        generate_moves
packages/sena_turo/builder/widgets/gps_chess_board.dart 458:67                    <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/sena_turo/builder/widgets/gps_chess_board.dart 380:53                    <fn>
packages/sena_core/src/editor/widgets/movable.dart 49:29                          <fn>
packages/flutter/src/widgets/drag_target.dart 517:18                              <fn>
packages/flutter/src/widgets/drag_target.dart 895:16                              finishDrag
packages/flutter/src/widgets/drag_target.dart 798:5                               end
packages/flutter/src/gestures/multidrag.dart 156:13                               [_up]
packages/flutter/src/gestures/multidrag.dart 258:12                               [_handleEvent]
packages/flutter/src/gestures/pointer_router.dart 98:12                           [_dispatch]
packages/flutter/src/gestures/pointer_router.dart 143:9                           <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/linked_hash_map.dart 21:13          forEach
packages/flutter/src/gestures/pointer_router.dart 141:17                          [_dispatchEventToRoutes]
packages/flutter/src/gestures/pointer_router.dart 127:7                           route
packages/flutter/src/gestures/binding.dart 488:19                                 handleEvent
packages/flutter/src/gestures/binding.dart 468:14                                 dispatchEvent
packages/flutter/src/rendering/binding.dart 333:11                                dispatchEvent
packages/flutter/src/gestures/binding.dart 413:7                                  [_handlePointerEventImmediately]
packages/flutter/src/gestures/binding.dart 376:5                                  handlePointerEvent
packages/flutter/src/gestures/binding.dart 323:7                                  [_flushPointerEventQueue]
packages/flutter/src/gestures/binding.dart 292:9                                  [_handlePointerDataPacket]
lib/_engine/engine/platform_dispatcher.dart 1319:13                               invoke1
lib/_engine/engine/platform_dispatcher.dart 303:5                                 invokeOnPointerDataPacket
lib/_engine/engine/pointer_binding.dart 168:39                                    [_onPointerData]
lib/_engine/engine/pointer_binding.dart 791:20                                    <fn>
lib/_engine/engine/pointer_binding.dart 720:14                                    <fn>
lib/_engine/engine/pointer_binding.dart 317:16                                    loggedHandler
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall
Error: RangeError (index): Index out of range: index must not be negative: -1
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 294:49      throw_
dart-sdk/lib/_internal/js_dev_runtime/private/js_array.dart 591:7                 _get]
packages/sena_core/src/pgn_parser/utils/chess.dart 1246:38                        generate_moves
packages/sena_turo/builder/widgets/gps_chess_board.dart 458:67                    <fn>
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54                runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 127:5                _async
packages/sena_turo/builder/widgets/gps_chess_board.dart 380:53                    <fn>
packages/sena_core/src/editor/widgets/movable.dart 49:29                          <fn>
packages/flutter/src/widgets/drag_target.dart 517:18                              <fn>
packages/flutter/src/widgets/drag_target.dart 895:16                              finishDrag
packages/flutter/src/widgets/drag_target.dart 798:5                               end
packages/flutter/src/gestures/multidrag.dart 156:13                               [_up]
packages/flutter/src/gestures/multidrag.dart 258:12                               [_handleEvent]
packages/flutter/src/gestures/pointer_router.dart 98:12                           [_dispatch]
packages/flutter/src/gestures/pointer_router.dart 143:9                           <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/linked_hash_map.dart 21:13          forEach
packages/flutter/src/gestures/pointer_router.dart 141:17                          [_dispatchEventToRoutes]
packages/flutter/src/gestures/pointer_router.dart 127:7                           route
packages/flutter/src/gestures/binding.dart 488:19                                 handleEvent
packages/flutter/src/gestures/binding.dart 468:14                                 dispatchEvent
packages/flutter/src/rendering/binding.dart 333:11                                dispatchEvent
packages/flutter/src/gestures/binding.dart 413:7                                  [_handlePointerEventImmediately]
packages/flutter/src/gestures/binding.dart 376:5                                  handlePointerEvent
packages/flutter/src/gestures/binding.dart 323:7                                  [_flushPointerEventQueue]
packages/flutter/src/gestures/binding.dart 292:9                                  [_handlePointerDataPacket]
lib/_engine/engine/platform_dispatcher.dart 1319:13                               invoke1
lib/_engine/engine/platform_dispatcher.dart 303:5                                 invokeOnPointerDataPacket
lib/_engine/engine/pointer_binding.dart 168:39                                    [_onPointerData]
lib/_engine/engine/pointer_binding.dart 791:20                                    <fn>
lib/_engine/engine/pointer_binding.dart 720:14                                    <fn>
lib/_engine/engine/pointer_binding.dart 317:16                                    loggedHandler
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 574:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 579:39  dcall
```

**Comment by alexobviously** [comment]
Hey, thanks for the report. I can confirm that I can recreate this, it's something to do with castling checks. Looking into it.

**Comment by alexobviously** [comment]
Okay actually in this case, what is happening is that bishop thinks the castling rights in this position are `KQkq`, when actually they are `-`. That's because the default behaviour is to fill in the missing castling rights with the full ones, but I think this is actually the wrong behaviour. I originally had it like this for early testing but coming back to it years later, I realise it's not really desirable. If your FEN is `k7/5p2/8/6P1/8/8/8/K7 b - - 0 1` then it works. However, I am going to change the default behaviour and put out a patch, so your FEN will work as expected.

**Comment by govind-maheshwari2** [comment]
Thank you
When can we aspect a patch release?

**Comment by alexobviously** [comment]
I'll release `1.4.1` in the next hour. Just rolling some other fixes in.

**Comment by govind-maheshwari2** [comment]
I also want to know it is possible in bishop to capture opponent piece using pawn by moving in straight instead of diagonal?

**Comment by alexobviously** [comment]
> I also want to know it is possible in bishop to capture opponent piece using pawn by moving in straight instead of diagonal?

Yeah, you can define almost anything that uses some combination of normal moves: what piece exactly do you want? If you just want a piece that moves forward and captures forward, with no double moves or en passant, just use `PieceType.fromBetza('fW')`. If you want a pawn that moves forward, and also has a double move on the first move, `fWifmnD` will work.

You can also do something like the [berolina pawn](https://en.wikipedia.org/wiki/Berolina_pawn) that moves diagonally and captures forwards.

There are lots of examples of piece types using betza notation in [piece_type.dart](https://github.com/alexobviously/bishop/blob/master/lib/src/piece_type.dart), but if you don't know betza and just want to know what to use, just lmk and I'll help.

**Comment by govind-maheshwari2** [comment]
Is there any way. I can contact you. I have some doubt and need some help regarding betza and introducing new piece using bishop.

**Comment by alexobviously** [comment]
Sure, you can make a discussion here: https://github.com/alexobviously/bishop/discussions

Gonna close this issue now since the bug is fixed in `1.4.1`
